[digiplex] Digiplex/EVO Binding initial contribution

rmichalak opened the pull request:
Provides integration with Digiplex/EVO alarm systems utilising PRT3 module (serial communication) as per issue #4526 

I finally decided to prepare PR as it appears I gained some supporters (https://community.openhab.org/t/digiplex-evo-alarm-system-integration-using-prt3-module/40409) and they all report that the binding actually works ;) .

davidgraeff commented on addons/binding/org.openhab.binding.digiplex/about.html:
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
```
Please use the NOTICE file instead of about.html (Eclipse license v2)

davidgraeff commented on addons/binding/org.openhab.binding.digiplex/pom.xml:
```diff
@@ -0,0 +1,18 @@
+	<artifactId>org.openhab.binding.digiplex</artifactId>
+
+	<name>Digiplex/EVO Binding</name>
+	<packaging>eclipse-plugin</packaging>
```
Please migrate to the new build-system, see the sticky issue for more information

davidgraeff commented on addons/binding/org.openhab.binding.digiplex/src/main/java/org/openhab/binding/digiplex/DigiplexBindingConstants.java:
```diff
@@ -0,0 +1,75 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
```
Please update the license headers. I thing the maven command is `mvn license:update` or similar.

davidgraeff commented on addons/binding/org.openhab.binding.digiplex/src/main/java/org/openhab/binding/digiplex/communication/AbstractResponse.java:
```diff
@@ -0,0 +1,35 @@
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * Abstract response
```
Can you add some javadoc, that explains the concepts and when to use this abstract class?

davidgraeff commented on addons/binding/org.openhab.binding.digiplex/src/main/java/org/openhab/binding/digiplex/communication/AbstractResponse.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ */
+@NonNullByDefault
+public abstract class AbstractResponse implements DigiplexResponse {
```
Make this an immutable class (no getters, all fields public final)

rmichalak replied:
It is an immutable class actually. I may add `final` to `success` field to make it more clear, but I don't see the reason to make it public and remove getters. 

If this is the way immutable classes are expressed in this project, then we should also modify all inheritors of `AbstractResponse` to keep the same pattern everywhere.

davidgraeff replied:
The classes are more concise, easier to review, and immutable class members cannot be changed anyway. Yes using the same pattern everywhere would reduce the submitted size by some extend and would be appreciated.

rmichalak replied:
OK, I can do that. 

I'm curious though is it a new direction going forward for immutable classes? Cause I can see many examples in the code that do not follow your guidelines and I'm a bit reluctant to introduce a new pattern, which looks more like an exception than a standard currently.

davidgraeff replied:
The old object oriented programming pattern requested to use getters. openHAB just turned 10 years, so yes not everything is consistent. 

In all my bindings and in all my binding reviews an immutable class is without getters.

davidgraeff commented on addons/binding/org.openhab.binding.digiplex/src/main/java/org/openhab/binding/digiplex/communication/AreaArmDisarmResponse.java:
```diff
@@ -0,0 +1,45 @@
+        this.type = type;
+    }
+
+    public static AreaArmDisarmResponse failure(int areaNo, ArmDisarmType type) {
```
Please add javadoc for public methods especially if they follow the builder pattern

davidgraeff commented on addons/binding/org.openhab.binding.digiplex/src/main/java/org/openhab/binding/digiplex/communication/AreaArmRequest.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Command for arming area
+ *
+ * @author Robert Michalak
```
Please check your files for the correct author tag usage pattern that is: "name - Initial contribution"

davidgraeff commented on addons/binding/org.openhab.binding.digiplex/src/main/java/org/openhab/binding/digiplex/discovery/DigiplexDiscoveryService.java:
```diff
@@ -0,0 +1,148 @@
+ *
+ */
+@NonNullByDefault
+public class DigiplexDiscoveryService extends AbstractDiscoveryService
```
Please make this a real OSGi service (`@Component`) with `immediate=false`.
There is a new pattern to be used for discovery services that are bound to handlers:
1. Return the class in your handlers `getServices()`.
2. Implement the required interface

rmichalak replied:
Do you have any example of such a usage? 

My understanding is that I should no longer extend `AbstractDiscoveryService` but implement `ThingHandlerService`, so that means I should no longer depend on `startScan` and `stopScan` methods from the former class but on `activate` and `deactivate` methods from the latter one?

davidgraeff replied:
You extend abstract discovery service and you implement the mentioned interface.

Unfortunately there is no openHAB documentation written yet, it is only mentioned on the old eclipse smarthome developer section.

Have a look for that interface here in the repository via the GitHub search. I will also look for one, just not now.

rmichalak replied:
That was my first impression as well, but unfortunately I received this error when doing so:

`The inherited method AbstractDiscoveryService.deactivate() cannot hide the public abstract method in ThingHandlerService`

I was able to overcome this by overriding `decativate` method to artificially change its visibility to `public`, but it's not a very elegant long term solution (to require it for every discovery service bound to a Thing).

There is one more missing piece in this area: what to do with `DigiplexHandlerFactory#registerDigiplexDiscoveryService` method? Is it still needed? My understanding of framework changes is that discovery services are now registered/removed automatically(?)

davidgraeff replied:
> I was able to overcome this by overriding decativate method to artificially change its visibility to public, but it's not a very elegant long term solution (to require it for every discovery service bound to a Thing).

Actually that is a requirement for all newly written services to expose activate/deactivate. The underlying/inherited ones are hidden from the OSGi framework. 

> Is it still needed? 

No. Can be removed. That is automatically handled.

rmichalak commented:
Hi @davidgraeff,

I think I addressed all your comments. I pushed another commit to make it easy for you to review the changes. As I understand the process, I will still need to squash these commits before they can find their way to master branch? Let me know when should I do it.

Regards,
Robert

davidgraeff commented:
The squashing part has been removed from the contribution readme file of this repo a few days ago, so all good. 

I don't know about the openHAB docs, might need to be changed as well.

I will do a second review when Travis turned green, please check why it fails.

Hilbrand commented:
I thought we had agreed at least 2 reviewers should approve a new binding? This binding has some fundamental flaws, like none of the classes are in an internal package.

davidgraeff commented:
Yup. I'd say I'm responsible to fix that then :)

This binding hasn't been reviewed for 3 months and the code quality is overall code and doesn't deserve to rot here forever like about 100 other PRs are doing it. 

But SAT would help a lot for new binding contributions. Would be cool if we could make that work again.

rmichalak commented:
Huh, Didn't want to create problems. 
The code is really ancient (written almost a year before the PR was created) and just adjusted to fit into a new project structure. I obviously missed new coding guidelines.
Happy to help and adjust the code as requested.

davidgraeff commented:
Having everything in `internal` was always part of the coding guideline (back in the days there was an exception for Thing handlers). I haven't noticed this, because I didn't even think about someone not doing that.
@rmichalak Can you move everything into an `internal` package? That would help. Otherwise I'm doing it later tonight.

rmichalak commented:
Well I can, just not today. If it is so urgent, then I would be happy to accept your offer. And again, sorry for causing problems.

wborn commented:
> I thought we had agreed at least 2 reviewers should approve a new binding?

Yes the rule is there for a reason. I wouldn't mind if we just revert commits for new add-ons that haven't been reviewed by two maintainers. WDYT @openhab/2-x-add-ons-maintainers ?

davidgraeff commented:
The Linux kernel has a staging repo as an alternative solution. People can already try bindings (the staging repo would upload its karaf features as well), the code evolves, can be statically checked periodically and fundamental changes don't come as a surprise to a contributor. If the contributor loses interest before a merge happens the binding would be removed from staging again.

Too many people in the maintainer team are only seeing the technical aspect. The psychologic effect for contributors if a new binding PR is neglected for 3 months and longer is fatal and everyone is doing it in free time.

There is also no alternative to this repo, as the eclipse marketplace will be shut down.

That's in no way encouraging to develop for oh. I rather say SAT should be improved to catch those mistakes. The reviewer should concentrate on the logical aspect, typical oh patterns and architecture and estimate if the addition is maintainable in the long run. 

One review would be enough under those conditions as a new binding would be guaranteed to not break anything else in the repo.

martinvw commented on bundles/org.openhab.binding.digiplex/src/main/resources/ESH-INF/i18n/digiplex_xx_XX.properties:
```diff
@@ -0,0 +1,13 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
```
This file should be removed

martinvw commented on bundles/org.openhab.binding.digiplex/src/main/resources/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,67 @@
+    xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+    xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
This label seems wrong

martinvw commented on bundles/org.openhab.binding.digiplex/src/main/resources/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,67 @@
+        </channels>
+
+		<config-description>
+		    
```
Formatting shows a mix of tabs and spaces might break the SAT when other users are running it.

rmichalak commented:
Ok, will try to fix everything this weekend. One question though: this PR is already closed, same as the corresponding issue #4526. Should I raise another issue for this or the current one will be reopened?

davidgraeff commented:
Open another issue and reference to this one

rmichalak commented:
@martinvw xml formatting was wrong indeed. 
All other issues (internal package, unused i18n file etc.) seems to be already fixed by @davidgraeff
Do you have any more comments on this binding, so I can fix them in bulk?
I raised #5422 for this purpose.

rmichalak commented:
I created #5428 PR. Can someone review?
